### PR TITLE
Introduces hack to pass variabels by reference

### PR DESCRIPTION
**Pull request by heiglandreas**
When mocking system-functions that expect variables to be passed by reference, this hack allows to do so. This might f.i. help mocking sort-functions or ```exec``` or ```passthru```

### CODE REVIEW

**Comment by heiglandreas** [comment]
Hold that one please. There is still stuff missing....
Gets a bit complicated.....

**Comment by malkusch** [comment]
Good one with the references. I missed that. Would you mind to add tests for the contribution as well. I'm waiting with the pull.

**Comment by heiglandreas** [comment]
I missed some details on the references so I'ts a bit harder than I originally thought. Currently I'm thinking on checking the original function and pass variables that are references in the original as references to the mock also. But that might take a day or two until I'm done with that one ;-)

Basically I want to mock ```exec``` and ```passthru``` and therefore I need the references...

And of course I'll then also add some tests ;-) 

I'll ping you when I'm through!

**Comment by malkusch** [comment]
Well, this is open source. Relax, there's no dead line. Thank you for your time contributing.

**Comment by heiglandreas** [comment]
That should be it. :smiley: 

:cry: - CodeCoverage dropped slightly in PHP5.6 as checking for ```...``` as parameter name will not work any more due to the new variadic functions but checking is needed for prior versions.

I hope I've covered all cases. I'd love to hear your feedback! :wink:

**Comment by malkusch** [comment]
I will go trough that PR slowly when my time allows that. Expect some comments on the code coming in.

**Comment by heiglandreas** [comment]
Great! :smiley: Take your time! It's a lot of stuff ;-)

File: classes/Mock.php [hunk 1]
```diff
@@ -137,7 +139,13 @@ public static function disableAll()
     public function call(array $arguments)
     {
         $this->recorder->record($arguments);
-        return call_user_func_array($this->function, $arguments);
+        
+        $Args = array();
```
Comment (malkusch) [inline classes/Mock.php]: Could you please change that to lower case (`$args`)? I know that this is not part of PSR-2, but I'm so much used to the Java convention.

File: classes/MockCallHelper.php [hunk 2]
```diff
@@ -23,14 +23,16 @@ class MockCallHelper
...
-        $canonicalFunctionName,
-        array $arguments
-    ) {
+    public static function call($functionName, $canonicalFunctionName, &$arguments)
+    {
         $registry = MockRegistry::getInstance();
         $mock     = $registry->getMock($canonicalFunctionName);
 
+        foreach ($arguments as $key => $arg) {
+            if ($arg === 'optionalParameter') {
```
Comment (malkusch) [inline classes/MockCallHelper.php]: `'optionalParameter'` will have to become a constant.

File: classes/Mock.php [hunk 3]
```diff
@@ -188,15 +196,84 @@ public function define()
...
+                return '';
+            }
+            $argument = '';
+            if (true === $arg->isPassedByReference()) {
+                $argument .= '&';
+            }
+            $argument .= '$' . $arg->name;
+
+            if ($arg->isOptional()) {
+                $argument .= ' = \'optionalParameter\'';
```
Comment (malkusch) [inline classes/Mock.php]: 'optionalParameter' will have to become a constant.

File: tests/MockTest.php [hunk 4]
```diff
@@ -125,6 +131,82 @@ public function testDisableAll()
         Mock::disableAll();
 
         $this->assertNotEquals(1234, time());
-        $this->assertEquals(1, min(1, 2));
+        $this->assertEquals(1, min(array(1, 2)));
```
Comment (malkusch) [inline tests/MockTest.php]: Semantically that's the same test. I'm just wondering why you did change that.

File: tests/MockTest.php [hunk 5]
```diff
@@ -64,8 +64,14 @@ public function testRecording()
...
-        $this->assertEquals(array(array(), array(true)), $recorder->getCalls());
+        $mock = new Mock(__NAMESPACE__, "abs", 'emptyFunc');
+        $mock->enable();
+        $recorder = $mock->getRecorder();
+        $this->assertEmpty($recorder->getCalls());
+
+        abs(12);
+        $this->assertEquals(array(array(12)), $recorder->getCalls());
+
     }
```
Comment (malkusch) [inline tests/MockTest.php]: I'm curious why you did change that test.

File: tests/global_functions.php [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+/**
+ * @author  Andreas Heigl<andreas@heigl.org>
+ * @license http://www.wtfpl.net/txt/copying/ WTFPL
+ * @since   08.02.15
+ */
+function emptyFunc()
+{
+    // Empty Function as required by codesniffer instead of an empty anonymous
+    // function
+}
```
Comment (malkusch) [inline tests/global_functions.php]: I don't think it is really necessary to bootstrap the tests with this function. A `function () {}` goes perfectly through my phpcs. For yours it doesn't? Which version do you use?

File: tests/phpunit/MockDisablerTest.php [hunk 7]
```diff
@@ -24,11 +24,11 @@ public function testEndTest()
...
         $min = new Mock(__NAMESPACE__, "min", "max");
         $min->enable();
-        $this->assertEquals(9, min(1, 9));
+        $this->assertEquals(9, min(array(1, 9)));
         
         $disabler = new MockDisabler($min);
         $disabler->endTest($this, 1);
         
-        $this->assertEquals(1, min(1, 9));
+        $this->assertEquals(1, min(array(1, 9)));
```
Comment (malkusch) [inline tests/phpunit/MockDisablerTest.php]: Again, curios why the change.

File: tests/phpunit/MockDisablerTest.php [hunk 8]
```diff
@@ -24,11 +24,11 @@ public function testEndTest()
     {
         $min = new Mock(__NAMESPACE__, "min", "max");
         $min->enable();
-        $this->assertEquals(9, min(1, 9));
+        $this->assertEquals(9, min(array(1, 9)));
```
Comment (malkusch) [inline tests/phpunit/MockDisablerTest.php]: Again, curios why the change.

File: classes/Mock.php [hunk 9]
```diff
@@ -188,15 +196,84 @@ public function define()
...
+    private function getArgumentsList()
+    {
+        $functionReflection = new ReflectionFunction('\\' . $this->name);
+        $argsReflection = $functionReflection->getParameters();
+        $arguments = array();
+
+        foreach ($argsReflection as $arg) {
+            if ($arg->name == '...') {
+                // If '...' is set as name of a parameter this is a variadic
+                // C-implementation before PHP5.5 - There is no way of knowing
```
Comment (malkusch) [inline classes/Mock.php]: I'm a bit outdated in PHP. Give me a hint what you mean by "this is a variadic C-implementation before PHP5.5". I understand that [`...`](http://php.net/manual/en/functions.arguments.php#functions.variable-arg-list.new) came with PHP-5.6. What is this variadic C-implementation before PHP-5.5?

File: tests/MockTest.php (hunk 5 above)
Comment (heiglandreas) [inline tests/MockTest.php]: I'm not actually sure why I removed that second ```time``-test. Seemd logical at the time of doing so....

I assume it was due to the fact that the test always failed due to the fact that the parameter to time is not necessary and therefore not added to the recording-array. Therefore the assertion would have to be rewritten to ```$this->assertEquals(array(array(), array()), $recorder->getCalls());``` where I assumed that it is a simple repetition of the previous assertion. But it also tests that **both** calls to the function are present in the recorder which is no not tested any more.

So I'll add that one with the "corrected" assertion.

File: tests/MockTest.php (hunk 4 above)
Comment (heiglandreas) [inline tests/MockTest.php]: The function signature changed between 5.5 and 5.6. In PHP5.5 the Reflection-API gave back 3 parameters for the ```min```-Function (```$param1```, ```$param2``` and ```...```) whereas in 5.6 that changed to 1 parameter (```array $array```) - In 5.6 the method also accepts the 5.5 way and vice versa. But as I used the Reflections function signature to get the parameters I had to change that test to work in PHP 5.6. 

I currently do not have a clue on how to change that while preserving the possibility to pass parameters by reference.

File: tests/global_functions.php (hunk 6 above)
Comment (heiglandreas) [inline tests/global_functions.php]: I'm using PHPCS 2.2.0. I did a ```composer require squizlabs/php_codesniffer``` to install codesniffer and then ran ```./vendor/bin/phpcs --standard=PSR2 classes/ tests/```. That failed (with a warning) due to the empty function in ```./tests/MockTest:172``` where I had originally ```$mock = new Mock(__NAMESPACE__, $name, function(){});```.

File: tests/phpunit/MockDisablerTest.php (hunk 8 above)
Comment (heiglandreas) [inline tests/phpunit/MockDisablerTest.php]: See above :wink:

File: tests/phpunit/MockDisablerTest.php (hunk 7 above)
Comment (heiglandreas) [inline tests/phpunit/MockDisablerTest.php]: See above :wink:

**Comment by malkusch** [comment]
Hi, would you mind squash everything to one commit. I would merge it then. Thank you very much.

**Comment by heiglandreas** [comment]
Great! THX! :smile:
